### ISSUE TITLE
Idea: Replace `dependent: :delete`/`dependent: :nullify` with FK `on_delete:`/`on_update:`

### ISSUE THREAD

**Issue by pirj** [djezzzl/database_consistency#91]
Hey @djezzzl ! :wave: 

Cascade deletion is sometimes a very expensive operation.
In case the association is defined as `dependent: :delete`, it is possible to rely on the database to perform the deletion instead.
See [`:on_delete`/`:on_update`](https://api.rubyonrails.org/v6.1.4/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_foreign_key).

Disclaimer: I clearly understand that `dependent: :destroy` should remain ignored, as [`destroy` also runs callbacks](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-destroy).

Would it make sense to detect this and suggest changing the FK in such a way that the DB handles cascade deletion/nullification?

Sorry for not being verbose. I'd love to validate the idea with you first, and elaborate if you support it.

Originally posted in the wrong repo: https://github.com/toptal/database_validations/issues/59

**Comment by pirj** [comment]
I'm up to send a PR with a new checker that would detect `dependent: :delete`/`dependent: :delete_all`/`dependent: :nullify` and suggest changing FKs to use `on_delete: :cascade`/`on_delete: :nullify` instead. Is this a good starting point? Should it be an `AssociationChecker` subclass?

Optional: recommend `on_delete: :restrict` instead of `dependent: :restrict_with_exception`.

> Rails skips validations on associated records

https://guides.rubyonrails.org/association_basics.html#options-for-has-many-validate
> If you set the :validate option to false, then associated objects will not be validated whenever you save this object. By default, this is true: associated objects will be validated when this object is saved.

I meant this. I guess it's only for the newly created associated records, though. Wondering how `validate: false` works:
```ruby
class School < ActiveRecord::Base
  has_many :students, validate: false
end

class Student < ActiveRecord::Base
  validate :name, presence: true
end

school.students << Student.new
school.save!
```
Won't that new student's `name` be validated for presence? 😱 

Additional reason to consider `on_delete` FK option:
> Note that :dependent is implemented using Rails' callback system, which works by processing callbacks in order. Therefore, other callbacks declared either before or after the :dependent option can affect what it does.

Related: https://rails.rubystyle.guide/#callbacks-order https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsactiverecordcallbacksorder (but it does not consider `dependent:` option.
Related: https://rails.rubystyle.guide/#before_destroy
> dependent => :destroy deletes children before "before_destroy" is executed

Choosing between delete and destroy: https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#module-ActiveRecord::Associations::ClassMethods-label-Delete+or+destroy-3F

**Comment by djezzzl** [comment]
> I'm up to send a PR with a new checker that would detect dependent: :delete/dependent: :delete_all/dependent: :nullify and suggest changing FKs to use on_delete: :cascade/on_delete: :nullify instead. Is this a good starting point?

Please do so :+1: 

> Should it be an AssociationChecker subclass?

I think yes. We check an association and it's option, if the `precheck` is `true` meaning it has `dependent:` option we:
- raise an error if there is no `on_delete` 
- raise another error if there is `on_delete`

> Optional: recommend on_delete: :restrict instead of dependent: :restrict_with_exception/dependent: :restrict_with_error.

I can understand why we need `on_delete` vs `delete` but this one I'm not yet sure, probably it should improve the consistency as callback is avoidable. Let's have it too then :+1: Could be part of the same checker. 

> Won't that new student's name be validated for presence? scream

Seems like that's the case, didn't check though.

**Comment by djezzzl** [comment]
Hi @pirj,

I have finally released this in 1.5.0. Please try it out and let me know if it works for you. 

Thank you for your suggestions! Keep it going, please!
